RealmSwift/SwiftUI.swift

```diff
@@ -645,6 +645,285 @@ extension EnvironmentValues {
+    @ObservedObject private var storage = ObservableAsyncOpenStorage()
+
+    private func asyncOpenForUser<T: BSON>(_ user: User, partitionValue: T, configuration: Realm.Configuration) {
+        let userConfig = user.configuration(partitionValue: partitionValue, cancelAsyncOpenOnNonFatalErrors: true)
+        let userSyncConfig = userConfig.syncConfiguration
+        var configuration = configuration
+        configuration.syncConfiguration = userSyncConfig
+        storage.asyncOpen(configuration: configuration)
+            .sink { completion in
+                if case .failure(let error) = completion {
+                    self.storage.asyncOpenState = .error(error)
+                }
+                cancel()
+            } receiveValue: { realm in
+                self.storage.realm = realm
+                self.storage.asyncOpenState = .open(realm)
+            }.store(in: &storage.cancellables)
+    }
+
+    /// :nodoc:
+    public var projectedValue: Published<AsyncOpenState>.Publisher {
+        return storage.$asyncOpenState
+    }
+
+    /// :nodoc:
+    public var wrappedValue: AsyncOpenState {
+        storage.asyncOpenState
+    }
+
+    /**
+     This will cancel any notification from the property wrapper states
+     */
+    public func cancel() {
+        storage.cancellables.forEach { $0.cancel() }
+        storage.cancellables = []
+    }
+
+    /**
+     Initialise the property wrapper
+     - parameter appId: The unique identifier of your Realm app.
+     - parameter partitionValue: The `BSON` value the Realm is partitioned on.
+     - parameter baseConfiguration: A Base Configuration use when creating the Realm,
+       if empty the configuraion is setted to the `defaultConfiguration`
+     - parameter timeout: The maximum number of milliseconds to allow for a connection to
+       become fully established., if empty or `nil` no connection timeout is setted.
+     */
+    public init<T: BSON>(appId: String, partitionValue: T, baseConfiguration: Realm.Configuration = Realm.Configuration.defaultConfiguration, timeout: UInt? = nil) {
+        let app = App(id: appId)
+        if let timeout = timeout {
+            let syncTimeoutOptions = SyncTimeoutOptions()
+            syncTimeoutOptions.connectTimeout = timeout
+            app.syncManager.timeoutOptions = syncTimeoutOptions
+        }
+        if app.currentUser?.isLoggedIn ?? false,
+           let currentUser = app.currentUser {
+            asyncOpenForUser(currentUser, partitionValue: partitionValue, configuration: baseConfiguration)
+        } else {
+            app.objectWillChange.sink { [self] app in
+                if let currentUser = app.currentUser {
+                    asyncOpenForUser(currentUser, partitionValue: partitionValue, configuration: baseConfiguration)
+                }
+            }.store(in: &storage.cancellables)
+        }
+    }
+
+    public mutating func update() {
+        if storage.realm == nil || storage.realm != realm {
```

Comment:
I believe we should be checking for configuration here, and also adding a Realm EnvironmentValue for partitionValue. There are use cases where partitionValues are derived and not static. This also should be tested.

Reply:
So I'm doing at some test for the `EnvironmentValue` for a partitionValue, the only issues with this is will be that we cannot have a generic environment value, but we can use AnyBSON, so the user would have to do something like this `.enviroment(.\partitionValue, .string("partitionValue"))` instead of been able to do this `enviroment(.\partitionValue, "partitionValue")` to inject it into the property wrapper

Reply:
In hindsight, making `BSON` equatable probably wasn't worth the hassle of having to use `AnyBSON` everywhere. Unfortunately, removing conformance would break semver, so there are two options: 

- Either we introduce a lower level protocol with no associated type that `BSON` conforms to, and then use that for places where PartitionValue makes sense:
```swift
protocol PartitionValue {}
protocol BSON: PartitionValue, Equatable {}
```
- Or we use AnyBSON as you suggested.
 
I attempted to get this working with opaque types, but sadly there seems to be a regression with opaque types that allows this to compile.